### PR TITLE
fix: suppress error notification for bugsnag

### PR DESCRIPTION
**Pull request by MoumitaM**
## PR Description

Please include a summary of the change along with the relevant motivation and context.

## Notion ticket

Ticket link

## Screenshots

Please add screenshots for any new features or UI bug fixes for the following browsers -

- Chrome
  >
- Firefox
  >
- Safari
  >

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: __tests__/utils/errorHandler.test.js [hunk 1]
```diff
@@ -36,3 +47,20 @@ describe("Test group for 'normalizeError' method", () => {
     expect(errMessage).toBe(`${staticMessage} "${JSON.stringify(obj)}"`);
   });
 });
+
+describe("Test group for 'handleError' method", () => {
```
Comment (bardisg) [inline __tests__/utils/errorHandler.test.js:51]: if we test after setTimeout lets add the done argument in test method and use it after expect

File: src/utils/errorHandler.js [hunk 2]
```diff
@@ -78,6 +78,9 @@ function handleError(error, customMessage, analyticsInstance) {
   }
 
   logger.error(errorMessage);
+  if(error.message && error.message.includes(FAILED_REQUEST_ERR_MSG_PREFIX)){
+    return; // do not notify errors in case the request has failed
```
Comment (bardisg) [inline src/utils/errorHandler.js:82]: can we make FAILED_REQUEST_ERR_MSG_PREFIX on array of messages to filter out like [FAILED_REQUEST_ERR_MSG_PREFIX] and iterate through all so we can filter more cases in the future as they come up?

File: src/utils/errorHandler.js [hunk 3]
```diff
@@ -1,21 +1,21 @@
...
  * This function is to send handled errors to available error reporting client
  *
  * @param {Error} error Error instance from handled error
  */
-function notifyError(error) {
+const notifyError = (error)=> {
   const errorReportingClient =
     window.rudderanalytics && window.rudderanalytics[ERROR_REPORTING_SERVICE_GLOBAL_KEY_NAME];
-
+
```
Comment (saikumarrs) [inline src/utils/errorHandler.js:12]:
```suggestion

```

File: src/utils/xhrModule.js [hunk 4]
```diff
@@ -79,16 +80,9 @@ class XHRQueue {
...
-              new Error(
-                `request failed with status: ${xhr.status}${xhr.statusText} for url: ${url}`,
-              ),
-            );
-            queueFn(
-              new Error(
-                `request failed with status: ${xhr.status}${xhr.statusText} for url: ${url}`,
-              ),
-            );
+            const errMessage = `${FAILED_REQUEST_ERR_MSG_PREFIX} ${xhr.status}${xhr.statusText} for url: ${url}`;
```
Comment (saikumarrs) [inline src/utils/xhrModule.js:83]:
```suggestion
            const errMessage = `${FAILED_REQUEST_ERR_MSG_PREFIX} "${xhr.status}" status text: "${xhr.statusText}" for URL: "${url}"`;
```

File: src/utils/xhrModule.js [hunk 5]
```diff
@@ -79,16 +80,9 @@ class XHRQueue {
...
+            handleError(new Error(errMessage));
+            queueFn(new Error(errMessage));
```
Comment (saikumarrs) [inline src/utils/xhrModule.js:85]:
```suggestion
            const err = new Error(errMessage);
            handleError(err);
            queueFn(err);
```

File: src/utils/constants.js [hunk 6]
```diff
@@ -47,6 +47,7 @@ const SYSTEM_KEYWORDS = ['library', 'consentManagement'];
 const UA_CH_LEVELS = ['none', 'default', 'full'];
 
 const DEFAULT_INTEGRATIONS_CONFIG = { All: true };
+const FAILED_REQUEST_ERR_MSG_PREFIX = 'request failed with status:';
```
Comment (saikumarrs) [inline src/utils/constants.js:50]:
```suggestion
const FAILED_REQUEST_ERR_MSG_PREFIX = 'Request failed with status:';
```

File: src/features/core/cookieConsent/OneTrust/browser.js [hunk 7]
```diff
@@ -6,9 +6,10 @@ class OneTrust {
     // If user does not load onetrust sdk before loading rudderstack sdk
     // we will not be filtering any of the destinations.
     if (!window.OneTrust || !window.OnetrustActiveGroups) {
-      throw new Error(
+      logger.error(
```
Comment (saikumarrs) [inline src/features/core/cookieConsent/OneTrust/browser.js:9]: This should be a warning message.

File: src/features/core/cookieConsent/OneTrust/browser.js [hunk 8]
```diff
@@ -6,9 +6,10 @@ class OneTrust {
...
         'OneTrust resources are not accessible. Thus all the destinations will be loaded',
       );
+      return;
```
Comment (saikumarrs) [inline src/features/core/cookieConsent/OneTrust/browser.js:12]: Returning from here alone may not be sufficient for the upward logic to work smoothly. Have you tested this scenario?

File: src/utils/errorHandler.js [hunk 9]
```diff
@@ -62,7 +62,7 @@ function normalizeError(error, customMessage, analyticsInstance) {
   return `[handleError]::${customErrMessagePrefix} "${errorMessage}"`;
 }
 
-function handleError(error, customMessage, analyticsInstance) {
+const handleError = (error, customMessage, analyticsInstance)=> {
```
Comment (saikumarrs) [inline src/utils/errorHandler.js:65]:
```suggestion
const handleError = (error, customMessage, analyticsInstance) => {
```

File: src/utils/errorHandler.js [hunk 10]
```diff
@@ -1,21 +1,21 @@
...
     window.rudderanalytics && window.rudderanalytics[ERROR_REPORTING_SERVICE_GLOBAL_KEY_NAME];
-
+    
   if (errorReportingClient && error instanceof Error) {
     errorReportingClient.notify(error);
   }
 }
 
-function normalizeError(error, customMessage, analyticsInstance) {
+const normalizeError = (error, customMessage, analyticsInstance)=> {
```
Comment (saikumarrs) [inline src/utils/errorHandler.js:18]:
```suggestion
const normalizeError = (error, customMessage, analyticsInstance) => {
```

File: src/utils/errorHandler.js [hunk 11]
```diff
@@ -1,21 +1,21 @@
...
-import { LOAD_ORIGIN, ERROR_REPORTING_SERVICE_GLOBAL_KEY_NAME } from './constants';
+import { LOAD_ORIGIN, ERROR_REPORTING_SERVICE_GLOBAL_KEY_NAME, FAILED_REQUEST_ERR_MSG_PREFIX } from './constants';
 
 /**
  * This function is to send handled errors to available error reporting client
  *
  * @param {Error} error Error instance from handled error
  */
-function notifyError(error) {
+const notifyError = (error)=> {
```
Comment (saikumarrs) [inline src/utils/errorHandler.js:9]:
```suggestion
const notifyError = (error) => {
```

File: src/utils/errorHandler.js (hunk 2 above)
Comment (saikumarrs) [reply src/utils/errorHandler.js:82]: @MoumitaM, do we have to handle the error in the XHR case? Won't it be sufficient to log the error message there itself?

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: src/features/core/cookieConsent/OneTrust/browser.js (hunk 8 above)
Comment (MoumitaM) [reply src/features/core/cookieConsent/OneTrust/browser.js:12]: I did but I will check again

File: src/features/core/cookieConsent/OneTrust/browser.js (hunk 7 above)
Comment (bardisg) [reply src/features/core/cookieConsent/OneTrust/browser.js:9]: But this assumes the user has loaded this integration as they were expecting OneTrust to work, feels like more of error to me

File: src/utils/errorHandler.js (hunk 2 above)
Comment (MoumitaM) [reply src/utils/errorHandler.js:97]: Yes, logging the error is sufficient. Now handleError method will do the same thing, only log the error.

Comment (saikumarrs) [reply src/utils/errorHandler.js:97]: No, I mean, we don't have to take it to the handleError method to just log the error.

File: __tests__/utils/errorHandler.test.js (hunk 1 above)
Comment (MoumitaM) [reply __tests__/utils/errorHandler.test.js:51]: Addressed

File: src/utils/errorHandler.js (hunk 2 above)
Comment (MoumitaM) [reply src/utils/errorHandler.js:97]: Addressed

File: src/features/core/cookieConsent/OneTrust/browser.js (hunk 8 above)
Comment (MoumitaM) [reply src/features/core/cookieConsent/OneTrust/browser.js:12]: Addressed

File: src/features/core/cookieConsent/OneTrust/browser.js (hunk 7 above)
Comment (saikumarrs) [reply src/features/core/cookieConsent/OneTrust/browser.js]: Makes sense. @MoumitaM, please revert the change.

File: src/features/core/cookieConsent/CookieConsentFactory.js [hunk 12]
```diff
@@ -17,8 +17,8 @@ class CookieConsentFactory {
       //     sourceConfig.cookieConsentManager &&
       // sourceConfig.cookieConsentManager.oneTrust &&
       // sourceConfig.cookieConsentManager.oneTrustenabled) {
-
-      return new OneTrust();
+      const oneTrust = new OneTrust();
+      return typeof oneTrust === 'object' && Object.keys(oneTrust).length > 0 ? oneTrust : null;
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply src/features/core/cookieConsent/CookieConsentFactory.js]: Addressed

File: src/features/core/cookieConsent/OneTrust/browser.js (hunk 7 above)
Comment (MoumitaM) [reply src/features/core/cookieConsent/OneTrust/browser.js:11]: Addressed

File: __tests__/utils/errorHandler.test.js (hunk 1 above)
Comment (MoumitaM) [reply __tests__/utils/errorHandler.test.js:51]: Addressed

File: src/utils/errorHandler.js [hunk 13]
```diff
@@ -60,9 +63,21 @@ const normalizeError = (error, customMessage, analyticsInstance)=> {
...
-}
+};
 
-const handleError = (error, customMessage, analyticsInstance)=> {
+/**
+ * A function to determine whether the error should be notified or not
+ * @param {Error} error
+ * @returns
+ */
+const shouldNotNotify = (error) => {
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply src/utils/errorHandler.js]: Addressed

File: src/utils/errorHandler.js [hunk 14]
```diff
@@ -93,10 +93,10 @@ const handleError = (error, customMessage, analyticsInstance) => {
   }
 
   logger.error(errorMessage);
-  if (shouldNotNotify(error)) {
-    return; // do not notify errors in case the request has failed
+  // do not notify errors in case the request has failed
```
Comment (saikumarrs) [inline src/utils/errorHandler.js:96]: The comment is not appropriate.

File: __tests__/utils/errorHandler.test.js (hunk 1 above)
Comment (saikumarrs) [reply __tests__/utils/errorHandler.test.js:51]: I don't think this is addressed.

File: src/features/core/cookieConsent/CookieConsentFactory.js [hunk 15]
```diff
@@ -1,7 +1,7 @@
 import { OneTrust } from './OneTrust';
 
-class CookieConsentFactory {
-  static initialize(cookieConsentOptions) {
+const CookieConsentFactory = {
```
Comment (saikumarrs) [inline src/features/core/cookieConsent/CookieConsentFactory.js:3]: We should revert this change.

File: src/features/core/cookieConsent/CookieConsentFactory.js [hunk 16]
```diff
@@ -17,13 +17,13 @@ class CookieConsentFactory {
       //     sourceConfig.cookieConsentManager &&
       // sourceConfig.cookieConsentManager.oneTrust &&
       // sourceConfig.cookieConsentManager.oneTrustenabled) {
-
-      return new OneTrust();
+      const oneTrust = new OneTrust();
+      return oneTrust.isInitialized ? oneTrust : null;
```
Comment (saikumarrs) [inline src/features/core/cookieConsent/CookieConsentFactory.js:21]: The initialized validation should happen within the APIs and return appropriate data.

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: src/features/core/cookieConsent/OneTrust/browser.js [hunk 17]
```diff
@@ -2,13 +2,16 @@ import logger from '../../../../utils/logUtil';
...
 class OneTrust {
+  isInitialized = false;
+
   constructor() {
     // If user does not load onetrust sdk before loading rudderstack sdk
     // we will not be filtering any of the destinations.
     if (!window.OneTrust || !window.OnetrustActiveGroups) {
-      throw new Error(
+      logger.error(
         'OneTrust resources are not accessible. Thus all the destinations will be loaded',
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply src/features/core/cookieConsent/OneTrust/browser.js]: Done

File: src/features/core/cookieConsent/CookieConsentFactory.js (hunk 16 above)
Comment (MoumitaM) [reply src/features/core/cookieConsent/CookieConsentFactory.js]: Done

File: src/features/core/cookieConsent/CookieConsentFactory.js (hunk 15 above)
Comment (MoumitaM) [reply src/features/core/cookieConsent/CookieConsentFactory.js:3]: Done

File: src/utils/errorHandler.js (hunk 14 above)
Comment (MoumitaM) [reply src/utils/errorHandler.js]: Done

File: __tests__/utils/errorHandler.test.js (hunk 1 above)
Comment (saikumarrs) [reply __tests__/utils/errorHandler.test.js:52]: This feels like a hack. Let's discuss this.

Comment (MoumitaM) [reply __tests__/utils/errorHandler.test.js:47]: Done

**Review by saikumarrs — APPROVED** [review]

**Review by bardisg — APPROVED** [review]
